### PR TITLE
cmake: build man pages

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
@cbodley would you mind taking a look?

File: man/CMakeLists.txt [hunk 1]
```diff
@@ -1,43 +1,60 @@
...
+
+foreach(manpage ${manpage_srcs})
+  string(REGEX MATCH "^[0-9]/" section ${manpage})
+  get_filename_component(cmd ${manpage} NAME_WE)
+  list(APPEND sphinx_input ${CMAKE_SOURCE_DIR}/doc/man/${manpage})
+  list(APPEND sphinx_output ${CMAKE_BINARY_DIR}/doc/man/${cmd}.${section})
+endforeach()
+
+install(FILES ${sphinx_output} ceph_selinux.8
+  DESTINATION ${CEPH_MAN_DIR}/man8)
```
Comment (tchaikov) [inline man/CMakeLists.txt]: we could use more sophisticated way to determine the section instead hard-wiring it to man8. but since we only *.8 man pages at this moment, let install all of them into "man8" now.

File: man/CMakeLists.txt [hunk 2]
```diff
@@ -1,43 +1,60 @@
...
-	rados.8
-	librados-config.8
-	ceph-clsinfo.8
-	ceph-debugpack.8
-	cephfs.8
-	ceph-dencoder.8
-	ceph-rest-api.8
-	ceph-post-file.8
-  DESTINATION ${CEPH_MAN_DIR}/man8)
+set(manpages_srcs
```
Comment (tchaikov) [inline man/CMakeLists.txt]: we list sphinx-build as one of the build dependencies, so i put the generated man pages into the build directory instead of in "/man".

also it might be more natural if we have a CMakeLists.txt in into "/doc/man/8", and collect the man pages there. but i am good either way.

File: CMakeLists.txt [hunk 3]
```diff
@@ -364,4 +373,4 @@ add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND})
 add_subdirectory(src)
 
 # man pages must be preprocessed, not supported yet
-#add_subdirectory(man)
+add_subdirectory(man)
```
Comment (cbodley) [inline CMakeLists.txt]: make this conditional on `WITH_MANPAGE`

File: man/CMakeLists.txt (hunk 2 above)
Comment (cbodley) [inline man/CMakeLists.txt]: > not sure we will have a "make dist" for cmake

that's my understanding as well. i don't think we want to involve cpack, either

> have a CMakeLists.txt in into "/doc/man/8", and collect the man pages there

i would prefer it this way too. for one, it would mean that the entries in `manpages_srcs` would be relative to their CMAKE_CURRENT_SOURCE_DIR, rather than having to add the `${CMAKE_SOURCE_DIR}/doc/man/` prefix. it also solves the install to "man8" piece because each CMakeLists.txt could do its own install(). but i'm fine with putting this off until we need it

**Comment by cbodley** [comment]
Thanks for taking this on @tchaikov, you're my hero! The install() part looks correct to me, but in testing 'make install' I'm getting the strangest errors:
```
-- Installing: /home/cbodley/local/share/man/man8
-- Up-to-date: /home/cbodley/local/share/man/man8
-- Up-to-date: /home/cbodley/local/share/man/man8
-- Up-to-date: /home/cbodley/local/share/man/man8
-- Up-to-date: /home/cbodley/local/share/man/man8
-- Up-to-date: /home/cbodley/local/share/man/man8
-- Up-to-date: /home/cbodley/local/share/man/man8
-- Up-to-date: /home/cbodley/local/share/man/man8
-- Installing: /home/cbodley/local/share/man/man8/ceph_selinux.8
CMake Error at man/cmake_install.cmake:36 (file):
  file INSTALL cannot copy file "/home/cbodley/ceph/man/ceph_selinux.8" to
  "/home/cbodley/local/share/man/man8/ceph_selinux.8".
Call Stack (most recent call first):
  cmake_install.cmake:38 (include)
```
The resulting directory looks like filesystem corruption:
```
$ ll ~/local/share/man/man8
ls: cannot access /home/cbodley/local/share/man/man8/ceph-fuse.8: Permission denied
total 0
?????????? ? ? ? ?            ? ceph-fuse.8
```
..but I was able to reproduce when installing to different devices, and @alimaredia reproduced it on his machine as well.

Digging further, I added some debugging to print the `$sphinx_output` variable and saw this:
`/home/cbodley/ceph/build/doc/man/ceph-fuse.8/;/home/cbodley/ceph/build/doc/man/rbd-fuse.8/;/home/cbodley/ceph/build/doc/man/ceph-rbdnamer.8/;/home/cbodley/ceph/build/doc/man/rbd-nbd.8/;/home/cbodley/ceph/build/doc/man/rbd-replay-prep.8/;/home/cbodley/ceph/build/doc/man/rbd-replay.8/;/home/cbodley/ceph/build/doc/man/rbdmap.8/;/home/cbodley/ceph/build/doc/man/rbd.8/`.

So it looks like we need to trim that trailing / off of the `section` variable that results from `string(REGEX MATCH "^[0-9]/" section ${manpage})`. There's probably a better way to do that, but adding `string(REPLACE "/" "" section ${section})` seemed to work for me and the man pages installed correctly.

File: man/CMakeLists.txt (hunk 2 above)
Comment (cbodley) [inline man/CMakeLists.txt]: typo on this line: manpages_srcs -> manpage_srcs

**Comment by cbodley** [comment]
@tchaikov I pushed fixes for the issues I mentioned, feel free to cherry-pick them from https://github.com/cbodley/ceph/commits/wip-cmake-manpages

**Comment by tchaikov** [comment]
@cbodley and @alimaredia thanks for testing this PR!

i merged your changes and restructured the CMakeLists.txt so we don't need to hardcode the "/doc/man" path.

File: man/CMakeLists.txt [hunk 4]
```diff
@@ -1,43 +1,4 @@
...
-if(WITH_RBD)
-  install(FILES
-	ceph-rbdnamer.8
-	rbd-nbd.8
-	rbd-replay-prep.8
-	rbd-replay.8
-	rbdmap.8
-	rbd.8
-	DESTINATION ${CEPH_MAN_DIR}/man8)
+if(WITH_SELINUX)
```
Comment (tchaikov) [inline man/CMakeLists.txt]: WITH_SELINUX is not supported yet in cmake, will leave it for another PR.

**Comment by cbodley** [comment]
the new stuff in doc/ looks good. retested and everything seemed to work :+1:

File: man/CMakeLists.txt (hunk 4 above)
Comment (tchaikov) [inline man/CMakeLists.txt]: added in #9133
